### PR TITLE
Add ultraplonk support

### PR DESCRIPTION
**Pull request by rushby**

### CODE REVIEW

File: tests/verify.test.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+describe('verify and subscribe to zkVerify events, wait for Attestation event then call proof of existence', () => {
+    const proofTypes = Object.keys(ProofType).map((key) => ProofType[key as keyof typeof ProofType]);
+
+    test.each(proofTypes)('should verify %s proof and respond on finalization with attestation confirmation then check proof of existence', async (proofType: ProofType) => {
+        console.log(`Running test for proof type: ${proofType}`);
+        const { proof, publicSignals } = loadProofData(ProofType.groth16);
+        const vk = loadVerificationKey(ProofType.groth16);
+
+        await performVerifyTransaction(ProofType.groth16, proof, publicSignals, vk, true, true);
```
Comment (yurikoinaba) [inline tests/verify.test.ts:27]:
```suggestion
        const { proof, publicSignals } = loadProofData(proofType);
        const vk = loadVerificationKey(proofType);

        await performVerifyTransaction(proofType, proof, publicSignals, vk, true, true);
```

File: tests/errors.test.ts [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
 import { zkVerifySession } from '../src';
-import { ZkVerifyEvents } from "../src/enums";
+import { ZkVerifyEvents } from "../src";
 
 jest.setTimeout(180000);
 
 describe('verify with bad data - Groth16', () => {
     it('should fail when sending groth16 data that cannot be formatted and emit an error event', async () => {
-        const dataPath = path.join(__dirname, 'data', 'groth16_error.json');
+        const dataPath = path.join(__dirname, 'common/data', 'groth16_error.json');
```
Comment (yurikoinaba) [inline tests/errors.test.ts:10]: Also update in line 53

**Review by yurikoinaba — COMMENTED** [review]
Also fix the `SupportedNetworks` import in `src/api/connection/index.test.ts`

File: tests/common/data/ultraplonk.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "vk": "",
```
[the remark this answers is not in the document]
Comment (rushby) [reply tests/common/data/ultraplonk.json:2]: I think this one it comes from a bin file instead but will double check!

File: tests/errors.test.ts (hunk 2 above)
Comment (rushby) [reply tests/errors.test.ts:10]: Doh! Thanks yes this was originally just for groth16 but needs changing to be all proof types

File: tests/verify.test.ts (hunk 1 above)
Comment (rushby) [reply tests/verify.test.ts:27]: Yes, was originally just testing groth1) but needs updating to be all proof types thanks 😃

File: tests/common/eventHandlers.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        expect(eventData.txHash).toBeDefined();
+
+        if (expectedDataType === 'vkRegistration') {
+            expect(eventData.statementHash).toBeDefined();
+            expect(eventData.attestationId).toBeUndefined();
+            expect(eventData.leafDigest).toBeUndefined();
+        } else if (expectedDataType === 'verify') {
+            expect(eventData.attestationId).not.toBeNull();
+            expect(eventData.leafDigest).not.toBeNull();
+            expect(eventData.statementHash).toBeUndefined();
```
Comment (DanieleDiBenedetto) [inline tests/common/eventHandlers.ts:49]: The registered vk event emits simply H(vk) while the statementHash is something emitted by the ProofVerified event and it's H(prefix, H(vk), H(pubs)). It seems to me here that:
- If "vkRegistration", then statementHash must be undefined while an additional param, vkHash must be present
- If "verify", then statementHash must be defined. Keep in mind that statementHash and leafDigest are exactly the same thing in this context.

File: tests/common/eventHandlers.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        }
+    });
```
Comment (DanieleDiBenedetto) [inline tests/common/eventHandlers.ts:70]: Same comment as above

File: tests/common/eventHandlers.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    expectedProofType: string,
+    expectedDataType: 'verify' | 'vkRegistration',
+): EventResults => {
+    const eventResults: EventResults = {
+        includedInBlockEmitted: false,
+        finalizedEmitted: false,
+        errorEventEmitted: false,
+        attestationConfirmedEmitted: false,
+        attestationBeforeExpectedEmitted: false,
+        attestationMissedEmitted: false,
```
Comment (DanieleDiBenedetto) [inline tests/common/eventHandlers.ts:25]: What's the difference among these 3 ?

File: tests/common/utils.ts [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+export const loadVerificationKey = (proofType: ProofType): string => {
+    if (proofType === ProofType.ultraplonk) {
+        const vkPath = path.join(__dirname, 'data', 'ultraplonk_vk.bin');
+        return fs.readFileSync(vkPath).toString('hex');
+    } else {
+        const dataPath = path.join(__dirname, 'data', `${proofType}.json`);
+        const proofData: ProofData = JSON.parse(fs.readFileSync(dataPath, 'utf8'));
+        return proofData.vk!;
+    }
+};
```
Comment (DanieleDiBenedetto) [inline tests/common/utils.ts:38]: Mmm perhaps here I would keep it more generic: no assumptions on the format and file names. After merging https://github.com/HorizenLabs/ultraplonk_verifier/pull/7 we can probably deprecate Noir <= v0.3.0 and things will be easier. But let's keep it for the future, don't want to disrupt anything before hackatons

File: tests/common/utils.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    console.log(`${proofType} Closing session...`);
+    await session.close();
+
+    return { eventResults, transactionInfo };
+};
+
+
+export const performVKRegistrationAndVerification = async (
```
Comment (DanieleDiBenedetto) [inline tests/common/utils.ts:93]: Perhaps it could be worth to also have another method for performing vk registration only ?

File: tests/common/utils.ts [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+
+export const validateTransactionInfo = (
+    transactionInfo: TransactionInfo,
+    expectedProofType: string
+): void => {
+    expect(transactionInfo).toBeDefined();
+    expect(transactionInfo.blockHash).not.toBeNull();
+    expect(transactionInfo.proofType).toBe(expectedProofType);
+    expect(transactionInfo.status).toBe(TransactionStatus.Finalized);
```
Comment (DanieleDiBenedetto) [inline tests/common/utils.ts:130]: Perhaps it could be worth to move this check up

File: tests/common/utils.ts [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    expect(transactionInfo.txHash).toBeDefined();
+    expect(transactionInfo.extrinsicIndex).toBeDefined();
+    expect(transactionInfo.feeInfo).toBeDefined();
+    expect(transactionInfo.weightInfo).toBeDefined();
+    expect(transactionInfo.txClass).toBeDefined();
```
Comment (DanieleDiBenedetto) [inline tests/common/utils.ts:135]: Very minor: probably not worth to perform this checks as we can assume that (if the transaction didn't error) the underlying APIs are correct and will populate these fields

File: tests/verify.test.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+describe('Register a VK and and then verify using the VK hash - All Proof Types', () => {
+    const proofTypes = Object.keys(ProofType).map((key) => ProofType[key as keyof typeof ProofType]);
+
+    test.each(proofTypes)('should register and verify %s proof and respond on finalization', async (proofType: ProofType) => {
+        const { proof, publicSignals } = loadProofData(proofType);
+        const vk = loadVerificationKey(proofType);
+
+        await performVKRegistrationAndVerification(proofType, proof, publicSignals, vk);
+    });
+});
```
Comment (DanieleDiBenedetto) [inline tests/verify.test.ts:40]: These tests include only happy paths. Could we also test negative cases ?

**Review by DanieleDiBenedetto — CHANGES_REQUESTED** [review]

File: tests/common/eventHandlers.ts (hunk 6 above)
Comment (rushby) [reply tests/common/eventHandlers.ts:25]: If a user signs up to listen to the events and they have specified a specific attestation id to wait for, the the zkverifyjs package will inform the user if an attestation before the one they are waiting for has happened (and continue listening), if the attestation event they received is after their expected value they are waiting for, or whether they event they are waiting for has now occurred.

File: tests/common/utils.ts (hunk 7 above)
Comment (rushby) [reply tests/common/utils.ts:38]: These are just tests, the format doesn't really matter for the files, the tests will turn the data into the zkverifyjs packages required format

File: tests/common/utils.ts (hunk 8 above)
Comment (rushby) [reply tests/common/utils.ts:93]: These are just integration tests though that demonstrate the user flow. Adding a separate test to test the same thing again seems unnecessary?

File: tests/common/utils.ts (hunk 10 above)
Comment (rushby) [reply tests/common/utils.ts:135]: Just to be safe though, I don't like to assume :D

File: tests/common/eventHandlers.ts (hunk 6 above)
Comment (DanieleDiBenedetto) [reply tests/common/eventHandlers.ts:25]: Thanks. Do you think the attestationMissed and attestationBeforeExpected have some use case ?

File: tests/verify.test.ts (hunk 11 above)
Comment (DanieleDiBenedetto) [reply tests/verify.test.ts:40]: Yes thanks ! One test for failure on verifying proof could also be useful

File: tests/common/utils.ts (hunk 9 above)
Comment (rushby) [reply tests/common/utils.ts:130]: Up to where - this is called as the first call in both:

`export const validateVerifyTransactionInfo = (
    transactionInfo: VerifyTransactionInfo,
    expectedProofType: string,
    expectAttestation: boolean
): void => {
    validateTransactionInfo(transactionInfo, expectedProofType);`
    
 And:
 
`export const validateVKRegistrationTransactionInfo = (
    transactionInfo: VKRegistrationTransactionInfo,
    expectedProofType: string
): void => {
    validateTransactionInfo(transactionInfo, expectedProofType);`
    
Or am i misunderstanding?

File: tests/common/eventHandlers.ts (hunk 4 above)
Comment (rushby) [reply tests/common/eventHandlers.ts:49]: These tests are checking the format the zkverifyjs package is responding with and confirming the packages logic is correctly mapping data for the events received rather than testing the pure format returned by zkverify itself.

These tests pass - is there something incorrect in the logic?

File: tests/errors.test.ts (hunk 2 above)
Comment (rushby) [reply tests/errors.test.ts:10]: Fixed.

File: tests/common/eventHandlers.ts (hunk 4 above)
Comment (DanieleDiBenedetto) [reply tests/common/eventHandlers.ts:49]: In the second if, leafDigest and statementHash are the same field, so no reason for them to be different.
In the first if, statementHash shall be replaced with vkHash.
I understand that these are tests, but nevertheless should provide guidance for developers (us or third party) looking for guidance on how to use the package, so the most accurate we are, the better imho

File: tests/common/utils.ts (hunk 7 above)
Comment (DanieleDiBenedetto) [reply tests/common/utils.ts:38]: Got it. But since they might be used by developers as guidance, let's make sure to add a test flow later when supporting Noir >= 0.31.0

File: tests/common/eventHandlers.ts (hunk 4 above)
Comment (rushby) [reply tests/common/eventHandlers.ts:49]: I still dont understand, is the confusion around the interface types being used to represent this data?  Is the naming of the fields not appropriate?

`export interface TransactionInfo {
  blockHash: string;
  proofType: ProofType;
  status: TransactionStatus;
  txHash?: string;
  extrinsicIndex?: number;
  feeInfo?: {
    payer: string;
    actualFee: string;
    tip: string;
    paysFee: string;
  };
  weightInfo?: {
    refTime?: string;
    proofSize?: string;
  };
  txClass?: string;
}

export interface VerifyTransactionInfo extends TransactionInfo {
  attestationId: number | undefined;
  leafDigest: string | null;
  attestationConfirmed: boolean;
  attestationEvent?: AttestationEvent;
}

export interface VKRegistrationTransactionInfo extends TransactionInfo {
  statementHash?: string;
}`

Comment (rushby) [reply tests/common/eventHandlers.ts:49]: The test is checking the return types, statementHash should not be defined for verify because the type is of VerifyTransactionInfo. Its just a check to make sure that is the case.

Same for attestationId and leafDigest on the VKRegistrationTransactionInfo type, they should not exist.

File: tests/common/eventHandlers.ts (hunk 6 above)
Comment (rushby) [reply tests/common/eventHandlers.ts:25]: I assume there is a possibility that someone starts listening for events expecting a particular attestation id and they either miss it, or they receive an earlier attestation and have to wait for the next one.

Just informational mostly for the user so they can see what's happening

**Review by yurikoinaba — APPROVED** [review]

**Review by lgiussan — APPROVED** [review]
LGTM!
